# Goal: Predict if a customer will churn (Yes = 1, No = 0)

In [1]:
# Import Libraries and Load Data
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

- use both RandomForestClassifier and LogisticRegression
- Try multiple algorithms Compare their performance Pick the best model for final predictions

- ### Accuracy_score
  - accuracy_score(y_true, y_pred) = how many predictions were correct
  - Accuracy = (Correct Predictions) / (Total Predictions)

- ### Classification_report
  - classification_report(y_true, y_pred) = Gives a detailed breakdown of model performance
  - Precision = How many predicted churned were actually churned?
  - Recall = How many actual churned customers were identified?
  - F1-score = Harmonic mean of precision and recall
  - Support = Count of actual samples for each class (0 and 1)

- ### Confusion_matrix
  - confusion_matrix(y_true, y_pred) = A 2x2 table that shows: Exact counts of TP, FP, FN, TN
  ### Confusion Matrix Explained
|                      | **Predicted: No (0)** | **Predicted: Yes (1)** |
|----------------------|-----------------------|-------------------------|
| **Actual: No (0)**   | ✅ True Negative (TN)  | ❌ False Positive (FP)  |
| **Actual: Yes (1)**  | ❌ False Negative (FN) | ✅ True Positive (TP)   |



In [2]:
# Load your cleaned CSV
df = pd.read_csv(r"C:\Users\bharg\Desktop\TEMP\Marketing Analytics & Customer Churn Prediction\datasets\cleaned\purchase_behavior.csv")
df.head()

customer_id  total_spent  number_of_orders  days_since_last_purchase  \
0  CUST_07035       341.85                11                        21   
1  CUST_04919        65.11                11                        51   
2  CUST_06927       284.36                10                       123   
3  CUST_10099        94.17                10                        55   
4  CUST_07489       607.98                 9                       118   

   avg_order_value customer_segment  churned  
0            31.08     Medium Value        0  
1             5.92        Low Value        0  
2            28.44     Medium Value        0  
3             9.42        Low Value        0  
4            67.55       High Value        0

In [3]:
# Clean & Prepare Data
df = df.drop(columns=["customer_id", "customer_segment"])
# customer_id drop because it’s just a unique name or number
# customer_segment drop because it’s created using other features like total spend or order value

# Fill or drop missing values
df = df.dropna()

df.isnull().sum()

total_spent                 0
number_of_orders            0
days_since_last_purchase    0
avg_order_value             0
churned                     0
dtype: int64

In [4]:
# Define Features and Target
# We are preparing the data for machine learning by splitting it into: X , y
# X → Features (the input the model will learn from) it is used to train 
# y → Target (the output the model will try to predict) it is use to check model's output
X = df.drop("churned", axis=1) # In X contain (total_spent, number_of_orders, days_since_last_purchase, avg_order_value)
y = df["churned"] # In y contain (churned)

In [5]:
# Train-Test Split
X_train, X_test, y_train, y_test =  train_test_split(X, y, test_size = 0.3 , random_state = 42)
# test_size = 0.3 - 30% of the data will be used for testing & remaining 70% will be used to train the model.  (0.3 It's a standard choice)
# random_state = 42 - Machine learning splits data randomly (42 is commonly used by data scientists)

In [6]:
# Train Models
# Logistic Regression
lr = LogisticRegression(max_iter=1000) # Try up to 1000 times to get the best result
lr.fit(X_train, y_train)                # Train the model using training data
y_pred_lr = lr.predict(X_test)         # Predict churn on test data

In [7]:
# Random Forest
rf = RandomForestClassifier(n_estimators=100, random_state=42) # Make a Random Forest model using 100 trees
rf.fit(X_train, y_train)            # Train the model
y_pred_rf = rf.predict(X_test)       # Predict churn

In [8]:
# Evaluate Performance
print("Logistic Regression Report:")
print(classification_report(y_test, y_pred_lr))
print(confusion_matrix(y_test, y_pred_lr))

print("\nRandom Forest Report:")
print(classification_report(y_test, y_pred_rf))
print(confusion_matrix(y_test, y_pred_rf))

Logistic Regression Report:
              precision    recall  f1-score   support

           0       0.87      1.00      0.93      3105
           1       0.76      0.06      0.11       495

    accuracy                           0.87      3600
   macro avg       0.81      0.53      0.52      3600
weighted avg       0.85      0.87      0.82      3600

[[3096    9]
 [ 467   28]]

Random Forest Report:
              precision    recall  f1-score   support

           0       0.89      0.97      0.93      3105
           1       0.54      0.22      0.31       495

    accuracy                           0.87      3600
   macro avg       0.71      0.59      0.62      3600
weighted avg       0.84      0.87      0.84      3600

[[3014   91]
 [ 388  107]]


In [ ]:
# Random Forest is better at catching customers who are about to leave.
# This is more helpful for a business because they can take action and try to stop those customers from leaving.

In [9]:
# Show Accuracy
print("Logistic Regression Accuracy:", round(accuracy_score(y_test, y_pred_lr),3))
print("Random Forest Accuracy:", round(accuracy_score(y_test, y_pred_rf),3))


Logistic Regression Accuracy: 0.868
Random Forest Accuracy: 0.867


### Logistic Regression has slightly higher accuracy (86.8%) than Random Forest (86.7%).
- Random Forest is still better at finding customers because who are likely to leave (churn), which is more important for the business than just accuracy

In [ ]:
import matplotlib.pyplot as plt

importances = rf.feature_importances_     # Gets importance of each feature
features = X.columns                      # Gets the names of your features

plt.title("Feature Importance - Random Forest")   # Adds a title
plt.barh(features, importances)           # Draws a horizontal bar chart
plt.figure(figsize=(10, 5))               # Sets the size of the chart
plt.show()     

# This shows which features (like total_spent, days_since_last_purchase) are most important for predicting churn.
# The longer the bar, the more the model depends on that feature.

In [ ]:
# Predict for New Data
# Format: [total_spent, number_of_orders, days_since_last_purchase, avg_order_value]
new_data = [[450.00, 3, 10, 150.00]]
prediction = rf.predict(new_data)
print("Will the customer churn? = ", prediction[0]) 
# What the Output Means:
# 0 → Not Churned
# 1 → Churned

- Churned (1)
  - He customer has left or stopped buying.
  - The business might lose this customer.
- Not Churned (0)
  - The customer is still active and buying.
  - They are loyal (for now).